Webscraping optional and api (#138)
UBC-DSCI/introduction-to-datascience: 01-reading.Rmd

- define the following:
    - absolute file path
    - relative file path

>     - URL

Reviewer: Is the first time we see the acronym URL? If so, should we define it here, or should we do that in the main text? I am thinking here and in the main text...

Author: agree -- both!

- retrieve data from a database query and bring it into R using the `collect` function from the `dbplyr` package
- use `write_csv` to save a data frame to a `.csv` file

> - (*optional*) obtain data using APIs and web scraping

Reviewer: Is the first time we see the acronym API? If so, should we define it here, or should we do that in the main text? I am thinking here and in the main text...

Author: agree, both!

    - read HTML source code from a URL using the `rvest` package

>     - read data from Twitter using the `rtweet` package

Reviewer: ```suggestion
    - read data from the Twitter API using the `rtweet` package
```

>     - compare downloading tabular data from a plain text file (e.g. `.csv`), accessing data from an API, and scraping the HTML source code a website

Reviewer: ```suggestion
    - compare downloading tabular data from a plain text file (e.g. `.csv`), accessing data from an API, and scraping the HTML source code from a website
```

write_csv(no_official_lang_data, "data/no_official_languages.csv")
```

## Obtaining data from the web 

> > This section is not required reading for the remainder of the textbook.

Reviewer: ```suggestion
> *This section does not contain pre-requisite knowledge for the remainder of the topics in the textbook, and is included for those leaners interested in learning a little bit more about how to obtain different types of data from the web.*
```

Author: i think the original wording is a bit clearer to be honest. i wouldn't mind adding something like your "is included for those...." to explain who should read the section though. Thoughts?

Reviewer: OK how about this then:

This section is not required reading for the remainder of the textbook. It is included for those leaners interested in learning a little bit more about how to obtain different types of data from the web.

I want to add the second sentence as the first read a bit terse to me, and it left me wondering if it was not required reading, why is it here.

into your browser window, your browser asks the web server to give it some data, 
and then your browser translates that data into something you can see. If the website
shows you some information that you're interested in, you could *create* a data set for yourself
by copying and pasting that information into a file. This process of taking information
directly from what a website displays is called *web scraping* (or sometimes *screen scraping*). Now, 
of course, copying and pasting information manually is a painstaking and error-prone process, especially when there is a lot of
information to gather. So instead of asking your browser to translate the information that the
web server provides into something you can see, you can collect
that data programmatically---in the form of **h**yper**t**ext **m**arkup **l**anguage (HTML) code---and
process it to extract useful information.

This subsection will show you the basics of both web scraping
with the [`rvest` R package](https://github.com/hadley/rvest)
and using the [`rtweet` R package](https://github.com/ropensci/rtweet)
to access the [Twitter](www.twitter.com) API.

### Web scraping

> **HTML and CSS selectors**

Author: I'm going to edit the defn earlier so we don't need to define something in a subsection heading

After installing and using SelectorGadget as shown in the video, we
obtain the two CSS selectors `.housing` and `.result-price`; we can use
these to 
scrape information about the square footage and the rental price, respectively.
The selector gadget returns them to us as a comma separated list (here
`.housing , .result-price`), which is exactly the format we need to provide to
R if we are using more than one CSS selector.

**STOP! Are you allowed to scrape that website?**

Before scraping data from the web, you should always check whether or not
you are *allowed* to scrape it! There are two documents that are important
for this: the `robots.txt` file and the Terms of Service
document. Let's take a look at Craigslist's Terms of Service document:

*"You agree not to copy/collect CL content via robots, spiders, scripts, scrapers, crawlers, or any automated or manual equivalent (e.g., by hand)."*

source: https://www.craigslist.org/about/terms.of.use

> Want to learn more about the legalities of web scraping and crawling? Read
> this interesting blog post titled ["Web Scraping and Crawling Are Perfectly
> Legal, Right?" by Benoit

> > Bernard](https://benbernardblog.com/web-scraping-and-crawling-are-perfectly-legal-right/).

Author: no-op here for now per slack

So what to do now? Well, we *could* ask the owner of Craigslist for permission to scrape.
However, we are not likely to get a response, and even if we did they would not likely give us permission.
The more realistic answer is that we simply cannot scrape Craigslist. If we still want
to find data about rental prices in Vancouver, we must go elsewhere.

To continue learning how to scrape data from the web, let's instead
scrape data on the population of Canadian cities from Wikipedia.
We have checked the [Terms of Service document](https://foundation.wikimedia.org/wiki/Terms_of_Use/en),
and it does not mention that web scraping is disallowed. In the video below, 
we demonstrate how to use the SelectorGadget tool to get CSS Selectors 
from [Wikipedia's Canada](https://en.wikipedia.org/wiki/Canada) for a table that
contains city names and populations from the 2016 Canadian Census: 

<iframe width="840" height="473" src="https://www.youtube.com/embed/O9HKbdhqYzk" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

We find that we should use the following CSS selector list:
```

> "td:nth-child(5) , td:nth-child(7) , .infobox:nth-child(122) td:nth-child(1) , .infobox td:nth-child(3)"

Author: removed quote and put each selector on its own line for better display

page <- read_html("https://en.wikipedia.org/wiki/Canada")
```

The `read_html` function directly downloads the source code for the page at 
the URL you specify, just like your browser would if you navigated to that site. But 
instead of  displaying the website to you, the `read_html` function just returns 
the HTML source code itself, which we have
stored in the `page` variable. Next, we send the page object to the `html_nodes`
function, along with the CSS selectors we obtained from
the SelectorGadget tool. Make sure to surround the selectors with quotations; `html_nodes` expects that
argument to be a string. The `html_nodes` function then selects *nodes* from the HTML document that 
match the CSS selectors you specified.
A *node* is an HTML tag pair combined with the content stored between the tags. For our
CSS selector `td:nth-child(5)`, an example node that would be selected would
be: 
```html

> <td style="text-align:left;background:#f0f0f0;"><a href="/wiki/London,_Ontario" title="London, Ontario">London</a></td>

Author: agree

```

We store the result of the `html_nodes` function in the `population_nodes` variable.
```{r 01-select-nodes}

> population_nodes <- html_nodes(page, "td:nth-child(5) , td:nth-child(7) , .infobox:nth-child(122) td:nth-child(1) , .infobox td:nth-child(3)")
> head(population_nodes)

Reviewer: The original `population_nodes` line was too long (again worried for print), so the only way to shorten it AFAIK so it will not be cut off in print is to use `paste0`. I don't love this, but I don't know what else to do... Other options for breaking the code across lines adds a newline character to it. If we do this, we will need some sort of note explaining why `paste0` is there... Sigh... 

```suggestion
css_selector <- paste0("td:nth-child(5) , td:nth-child(7) , ",
                       ".infobox:nth-child(122) td:nth-child(1) , ",
                       ".infobox td:nth-child(3)")
population_nodes <- html_nodes(page, css_selector)
head(population_nodes)
```

Author: fixed with explanation

```

Next we extract the meaningful data---in other words, we get rid of the HTML code syntax and tags---from 
the nodes using the `html_text`
function. In the case of the example
node above, `html_text` function returns `"London"`.

```{r 01-get-text}
population_text <- html_text(population_nodes)
head(population_text)
```

Fantastic! We seem to have extracted the data of interest from the 
raw HTML source code. But we are not quite done; the data
is not yet in an optimal format for data analysis. Both the city names and
population are encoded as characters in a single vector, instead of being in a
data frame with one character column for city and one numeric column for
population (like a spreadsheet).
Additionally, the populations contain commas (not useful for programmatically
dealing with numbers), and some even contain a line break character at the end

> (`\n`). In the next chapter, we will learn more about how to *wrangle* this data
> into a more useful format for data analysis using R.

Reviewer: ```suggestion
(`\n`). In the next chapter, we will learn more about how to *wrangle* data
such as this into a more useful format for data analysis using R.
```

### Using an API

Rather than posting a data file at a URL for you to download, many websites these days
provide an API that must be accessed through a programming language like R. The benefit of this

> is that data owners have much more control over the data they provide to users. However, unlike

Author: I don't think we want to repeat the argument again, but rather provide a summarized version of it here (which is already done)

idea that you can learn how to use another API if needed.

In particular, in this book we will show you the basics of how to use
the [rtweet](https://github.com/ropensci/rtweet) package in R
to access data from the [Twitter](www.twitter.com) API. One nice feature of this particular 
API is that you don't need  a special *token* to access it; you simply need to 
make an account with them. Your access to the data will then be authenticated and controlled through
your account username and password. If you have a Twitter account already (or are willing to make one), you can follow
along with the examples that we show here. To get started, load the `rtweet` package:
```r
library(rtweet)
```
This package provides an extensive set of functions to search Twitter for tweets, users, their friends, and more. 
Let's construct a small dataset of the last 400 tweets from the Tidyverse account. A few of the most recent tweets
are shown below.

```{r 01-tidyverse-twitter, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "The Tidyverse Twitter feed. [Source](https://twitter.com/tidyverse?lang=en)", fig.retina = 2, out.width="800"}
knitr::include_graphics("img/tidyverse_twitter.png")
```

> **STOP! Think about your API usage carefully!**

Author: I just added a new bolded subsection header after the paragraph (to mimic the earlier stop section on web scraping).

> When you access an API, you are initiating a transfer of data from a web server

Author: we actually use that term a couple times earlier. any suggested alternative? all the ones i can think of are less clear and/or more technical.

> When you access an API, you are initiating a transfer of data from a web server

Author: resolved via commit

to your computer. Web servers are expensive to run and do not have infinite resources.
If you try to ask for *too much data* at once, you can use up a huge amount of the server's bandwidth. 

> If you try to ask for data *too frequently*---e.g. if you 

Author: we don't do this anywhere else (at least for now). should decide what to do later on when formatting

make many requests to the server in quick succession---you can also bog the server down and make
it unable to talk to anyone else. Most servers have mechanisms to revoke your access if you are not

> careful, but you should try to prevent issues from happening in the first place by being extra careful

Author: I prefer the original -- less repetitive

with it. Note that `rtweet` can be used to manage most aspects of your account (make posts, add friends, etc.),
which is why `rtweet` asks for such extensive permissions.
If you decide to allow `rtweet` to talk to the Twitter API using your account information, then 
input your username and password and hit "Sign In." Twitter will probably send you an email to say
that there was an unusual login attempt on your account, and in that case you will have to take
the one-time code they send you and provide that to the `rtweet` login page too. 

> Every API has its own way to authenticate users when they try to access data. Many APIs require you to
> sign up to receive a *token*, which is a secret password that you input into the R package (like `rtweet`) 
> that you are using to access the API. 

With the authentication setup out of the way, let's run the `get_timelines` function again to actually access
the API and take a look at what was returned:
```r
rt <- get_timelines('tidyverse', n=400)
rt
```
```{r 01-reading-hidden-load-tweets, echo = FALSE, message = FALSE, warning = FALSE}
rt <- read_csv("data/tweets.csv")
rt
```

> The data has quite a few variables! Let's reduce this down to a few variables of interest: `created_at`, `text`, `is_retweet`, and `retweet_screen_name`.

Reviewer: Changed order, to reflect my changed order in the code below.
```suggestion
The data has quite a few variables! Let's reduce this down to a few variables of interest: `created_at`,  `retweet_screen_name`, `is_retweet`, and `text`,.
```

```{r 01-select-tweets, message = FALSE, warning = FALSE}
rt <- rt %>% select(created_at, text, is_retweet, retweet_screen_name)
rt
```

> If you look back up at the image of the Tidyverse twitter page, you will

Reviewer: I think the format of data retrieved from this API is in stark contrast to the format of data retrieved from the web scraping example. We should make a comment here about this. In particular, the readers might think that API's always give back data frames. This is not true. So a comment stating that API's generally give you back more homogenous (or organized?) data with less code as compared to web scraping, is fair, I think. But I think we're also warranted to state that different data formats will be given by different API's, or even different functions within the same API!

Author: resolved via commit

recognize the text of the most recent few tweets in the above dataframe.  In
other words, we have successfully created a small dataset using the Twitter

> API---neat! From this point onward, the `rt` dataframe is stored on your

Reviewer: ```suggestion
API---neat! From this point onward, the `tidyverse_tweets` dataframe is stored on your
```

machine, and you can play with it to your heart's content; you can use

> `write_csv` to save it to a file, `read_csv` to read it into R again later, and

Reviewer: Could we suggest they can now do more interesting things than reading and writing data? They could calculate the percentage of tweets that are retweets compared to tweets? They could find the five Twitter screen names most often retweeted by the Tidyverse Twitter account? They could create visualizations for these. These are things they do not yet know how to do, but will learn in the next two chapters. So we could state that as well (encouragement to keep reading!).

Author: resolved via commit
